### PR TITLE
Branch 165019969

### PR DESCRIPTION
**Pull request by av8ramit**
Needed a CP for 1.3

### CODE REVIEW

**Comment by rmlarsen** [comment]
The device_setter_test failure looks like a real bug:

ERROR: testBasic (__main__.RandomStrategyTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/var/lib/jenkins/workspace/tensorflow-pull-requests-cpu-python3/bazel-ci_build-cache/.cache/bazel/_bazel_jenkins/eab0d61a99b6696edb3d2aff87b585e8/execroot/workspace/bazel-out/local-py3-opt/bin/bazel_pip/tensorflow/contrib/training/device_setter_test.runfiles/org_tensorflow/bazel_pip/tensorflow/contrib/training/python/training/device_setter_test.py", line 42, in testBasic
    u = variables.Variable(array_ops.zeros([2, 2]))
  File "/workspace/pip_test/venv/lib/python3.4/site-packages/tensorflow/python/ops/variables.py", line 208, in __init__
    constraint=constraint)
  File "/workspace/pip_test/venv/lib/python3.4/site-packages/tensorflow/python/ops/variables.py", line 326, in _init_from_args
    name=name)
  File "/workspace/pip_test/venv/lib/python3.4/site-packages/tensorflow/python/ops/state_ops.py", line 131, in variable_op_v2
    shared_name=shared_name)
  File "/workspace/pip_test/venv/lib/python3.4/site-packages/tensorflow/python/ops/gen_state_ops.py", line 910, in _variable_v2
    shared_name=shared_name, name=name)
  File "/workspace/pip_test/venv/lib/python3.4/site-packages/tensorflow/python/framework/op_def_library.py", line 787, in _apply_op_helper
    op_def=op_def)
  File "/workspace/pip_test/venv/lib/python3.4/site-packages/tensorflow/python/framework/ops.py", line 2669, in create_op
    self._apply_device_functions(ret)
  File "/workspace/pip_test/venv/lib/python3.4/site-packages/tensorflow/python/framework/ops.py", line 3438, in _apply_device_functions
    op._set_device(device_function(op))  # pylint: disable=protected-access
  File "/workspace/pip_test/venv/lib/python3.4/site-packages/tensorflow/python/training/device_setter.py", line 114, in device_function
    ps_device.task = self._ps_strategy(op)
  File "/workspace/pip_test/venv/lib/python3.4/site-packages/tensorflow/contrib/training/python/training/device_setter.py", line 51, in __call__
    n = int(hashlib.md5(key).hexdigest(), 16)
TypeError: Unicode-objects must be encoded before hashing

File: tensorflow/contrib/training/python/training/device_setter.py [hunk 1]
```diff
@@ -19,11 +19,39 @@
...
+    """Creates a new `RandomStrategy`."""
+    self._num_tasks = num_ps_tasks
+    self._seed = seed
+
+  def __call__(self, op):
+    """Chooses a ps task index for the given `Operation`."""
+    key = "%s_%d" % (op.name, self._seed)
+    # Use MD5 instead of Python's built-in hash() to get consistent outputs
+    # between runs.
+    n = int(hashlib.md5(key).hexdigest(), 16)
```
Comment (rmlarsen) [inline tensorflow/contrib/training/python/training/device_setter.py]: You can see an example of how to do it here:

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/rnn_cell_impl.py#L726

**Review by rmlarsen — CHANGES_REQUESTED** [review]

File: tensorflow/contrib/training/python/training/device_setter.py (hunk 1 above)
Comment (av8ramit) [inline tensorflow/contrib/training/python/training/device_setter.py]: Done. Thanks @rmlarsen

**Review by rmlarsen — APPROVED** [review]

**Comment by av8ramit** [comment]
Cmake has a hash mismatch which is weird, and metric_ops test is failing.

**Comment by rmlarsen** [comment]
Maybe just disable metric_ops and ask the author to fix it?
Let me look at the Cmake error. (I ignored it since it seems perpetually broken :-( )

**Comment by rmlarsen** [comment]
@av8ramit something is fishy. The hash of the mirror and the original do not match:

 http://mirror.bazel.build/github.com/NVlabs/cub/archive/1.6.4.zip  => 966d0c4f41e2bdc81aebf9ccfbf0baffaac5a74f00b826b06f4dee79b2bb8cee

 https://github.com/NVlabs/cub/archive/1.6.4.zip => b35162e70fdfee4af4926b46dc3fbb89e06c57bb5b2e72bf6cc47e1f58cc2c75

@jart @ebrevdo do you know how this could be?

**Comment by rmlarsen** [comment]
@av8ramit considering that the test catches the hash problem (whenever we fall back to the original) and it was already present before this push, I think it's fine to merge this.

**Comment by av8ramit** [comment]
@rmlarsen I updated the cmake hash to match the file it downloads. The mirror link and cmake link have diverged which is why regular bazel is fine.

**Comment by av8ramit** [comment]
Jenkins, test this please.

**Comment by av8ramit** [comment]
Jenkins, test this please.

**Comment by gunan** [comment]
Looks like we still have some build issues on GPU.
Maybe we should create a new push?

**Comment by av8ramit** [comment]
Yeah and I've deviated a couple commits now anyways. I'm fine closing and waiting for a new push.

**Pull request closed without merging**